### DISCUSSION TITLE
Allow exposing lower access level members without per-member annotation

### DISCUSSION THREAD

**Discussion by hedukas** [gohanlon/swift-memberwise-init-macro#36] [Ideas]
In my instance there are a lot of types in the codebase that have private properties which are then exposed through the init. And also sometimes a mixture of private and internal and public variables. (Example below)

```swift
public struct ViewModel {
    private let title: String
    private let subtitle: String?
	
    public init(title: String, subtitle: String?) {
        self.title = title
        self.subtitle = title
    }
}
```

I would love to use the macro for this, but it means that the usage looks like this

```swift
@MemberwiseInit(.public)
public struct ViewModel {
    @Init(.public)
    private let title: String
    @Init(.public)
    private let subtitle: String?
}
```

While it’s not too much trouble for such a small type, the more variables you have the more annoying it gets. At that point it seems easier (and less cluttered) to just write the initialiser yourself.

So I’m proposing adding a parameter to the macro that would allow you to specify what kind of access level variables should be included in the initialiser. So the usage could look like this:

```
@MemeberwiseInit(.public, withLowerAccessLevelMembers: true)
public struct ViewModel {
    public let title: String
    private let subtitle: String?

    public(title: String, subtitle: String?) {} // the generated init
}
```

This could make the macro even easier to use across different types. If this is acceptable, I’m happy to contribute the change.

**Comment by gohanlon** [comment]
Hi @hedukas, thank you for your thoughtful suggestion!

Personally, I'm torn. I seldomly use `@Init(.public)`:

- I'd usually make the member public. If I'm already choosing to expose it publicly via the `init`, making the member itself public isn't a large change. (Even more so for `let` properties.)
- Value semantics make access control less critical for API design, since you don't need to leverage access control to prevent mutation of shared references.

```swift
@MemberwiseInit(.public)
public struct ViewModel {
  public let title: String
  public let subtitle: String?
}
```

I think making members public is a better practice in general, and I'd almost always recommend it over `@Init(.public)`. Even if the proposed feature were added, I'd still recommend making members public instead.

> Aside: MemberwiseInit's README prominently recommends using `@Init(.public)`, but it really shouldn't.

Like Swift itself, I've made `@MemberwiseInit` safe by default. Yet Swift does provide plenty of foot guns that require explicit intent, e.g. `UnsafeMutablePointer`. Perhaps your proposal, with a stronger name of warning, could work. Something like: `unsafeIncludeAllMembers`.

Another concern is that this feature may be hard to get right. There would be several usage interactions that need to be carefully considered. A couple come to mind:

### `@Init(.internal)`

```swift
@MemberwiseInit(.public, unsafeIncludeAllMembers: true)
public struct ViewModel {
  public let title: String
  @Init(.internal) private let subtitle: String?
}
```

This should probably result in a either an error or warning diagnostic (and fix-its), since both `@Init(.internal)` and `unsafeIncludeAllMembers: true` are explicit intents that conflict. And, diagnostics and fix-its are hard to get right. (Currently, the [core logic](https://github.com/gohanlon/swift-memberwise-init-macro/blob/main/Sources/MemberwiseInitMacros/Macros/MemberwiseInitMacro.swift) of `@MembewiseInit` is ~400 lines, and the [diagnostics](https://github.com/gohanlon/swift-memberwise-init-macro/blob/main/Sources/MemberwiseInitMacros/Macros/Support/Diagnostics.swift) are ~500 complex lines.)

### `@Init(.ignore)`

```swift
@MemberwiseInit(.public, unsafeIncludeAllMembers: true)
public struct ViewModel {
  public let title: String
  @Init(.ignore) private let subtitle: String?
}
```

This shows a shortcoming of the name `unsafeIncludeAllMembers` — your more precise `withLowerAccessLevelMembers` would avoid this ambiguity. (Maybe `exposeLowerAccessLevelMembers`?)


## Closing

What do you think? Can you think of any other cases that need careful consideration? Or, is simply making members public a viable option for your structs?

**Comment by hedukas** [comment]
Hi! Thanks for the reply and sorry it took so long to get back. 

**Value semantics**

For context in our codebase we have a lot of frameworks. Sometimes we need to pass just the type, sometimes we need to initialise it, but we have taken the YAGNI approach. We’ll only make something public when it becomes unnecessary. 

For example we have 1500 public structs (about 5000 in total) in our project out of which over 300 have private properties which could all potentially benefit from this macro and even more if we start counting classes as well. It’s a very common pattern (at least for us). 

For a viewmodel yes it makes more sense to keep the members at the same level as the struct, but if we want to use it on more (for example in all viper stack components), then I can’t see the macro being useful without a flag such as this. Or is this much larger of a scope than this macro is made for?

**Conflicting macros**

`@Init(.internal)`

I see where you’re coming from and I indeed did not consider this before. If they were allowed to be used at the same time it could create a lot of confusion. This is where I think errors and diagnostic could help with. It could show an error that either the `@Init` or `withLowerAccessLevelMembers` should be removed. Would this help with the potential confusion here?

This would apply to all the different access levels this macro could have.

`@Init(.ignore)`

This flag I think should be respected when the `withLowerAccessLevelMembers` is also in use. Because it isn’t actually about the access level of a member but rather whether to include the property in the initializer or not, regardless of their access level.

Because of this I think it wouldn’t be too hard to assume how this ignore flag cooperates with `withLowerAccessLevelMembers` flag.

**Comment by nathansalaun-sofia** [comment]
I was about to create a thread to make the same proposal.

My app is split in multiple SPM packages to improve compilation times. I have structs that I use as "input" of the code in the packages, that the main app instantiates and gives when calling a function, but never reads any fields from. 

So the init is exposed and public but all fields are internal and unused by the importers of the package. This is the intended design of my packages API.

Right now I put all fields in `public` but I'd rather not have the macro interfere with the code, whether it is by adding `public` or `@Init(.public)` - I want it to generate the initializer matching the current code and that's it. I don't want to alter my struct to accommodate from the macro, which is supposed to make my code slimmer.

Maybe instead of a `withLowerAccessLevelMembers` parameter, add a "default @Init" access modifier parameter? Something like `@MemberwiseInit(.public, defaultInitVisibility: .public)` which would be equivalent to adding `@Init(.public)` to all properties?

**Comment by gohanlon** [comment]
How about adding a special `@UnsafeMemberwiseInit` macro? Here's my thinking:

- It's "unsafe" because it makes it easier to accidentally expose internal implementation details that were meant to be private, e.g. if a private field is added later without realizing it will be exposed by the generated init.
- It's "unsafe" because it goes against the principle of explicitly declaring what should be part of the public API.
- The name "unsafe" should discourage use, except by those who understand the tradeoffs.
- `@UnsafeMemberwiseInit(.public)` is succinct, e.g. for use with large number of `Codable` structs.

Example:

```swift
@UnsafeMemberwiseInit(.public)
public struct ViewModel {
  private let title: String
  private let subtitle: String?

  // Macro generates init:
  public init(title: String, subtitle: String?) {
    self.title = title
    self.subtitle = title
  }
}
```

Implementation-wise, this still needs careful consideration of edge cases and potential conflicts with other `@Init` declarations.

@nathansalaun-sofia @hedukas Thoughts?

**Reply by nathansalaun-sofia** [reply]
I think this is a great idea!

**Comment by hedukas** [comment]
While I like the idea of a separate macro for this I don’t really agree that it is unsafe. And accidentally exposing something would be caught by building the app very fast? I don't think these would make it to code you ship? And just by looking at it I think people would get confused by what exactly is unsafe. 

In addition I think it will create more friction when writing code. If I had a type that used this unsafe annotation and I update the type to not have private properties anymore then I’d also need to change the annotation (and know I need to change it). Though you could argue that the same applies if it had an property attached.

I also don’t think this is a behaviour we want to discourage. If we were to advocate for everything to be made the same access level as the type then we’d be exposing so much more to the entire codebase than what is necessary. In the codebase I’m working on we’re rather encourage encapsulation.

**Reply by gohanlon** [reply]
> [...] And accidentally exposing something would be caught by building the app very fast? I don't think these would make it to code you ship? [...]

Unfortunately, no — the compiler won't always catch this.

**Reply by nathansalaun-sofia** [reply]
How about `Unchecked` instead of `Unsafe` then?

**Reply by gohanlon** [reply]
I think `@UncheckedMemberwiseInit(.public)` is a much better name. It aligns with Swift's use of "unchecked" for bypassing compiler checks (e.g., `@unchecked Sendable`), more accurately implies the macro's behavior, and appropriately cautions developers without implying runtime safety issues.

**Comment by gohanlon** [comment]
@hedukas @nathansalaun-sofia Thank you for your valuable input and discussion on this issue. Based on our conversation, I've implemented `@_UncheckedMemberwiseInit` as an experimental feature, included in [0.5.0](https://github.com/gohanlon/swift-memberwise-init-macro/releases/tag/0.5.0).

Key points:
1. The macro includes all properties, regardless of access level.
2. It has reduced compile-time safety checks compared to `@MemberwiseInit`.
3. Unlike `@MemberwiseInit`, `@_UncheckedMemberwiseInit` includes attributed properties by default.
4. You can use `@Init(.ignore)` or `@InitWrapper(type:)` to explicitly handle attributed properties if needed.

I encourage you to try out this new macro and provide feedback on its usage. Your experiences will help us refine the feature and decide on its future stable implementation.

**Comment by nathansalaun-sofia** [comment]
@gohanlon the new macro works great for me so far, thanks!

What do you mean by `Unlike @MemberwiseInit, @_UncheckedMemberwiseInit includes attributed properties by default.`? Does that mean that properties with property wrappers or attached macros will be included in the generated init?

**Reply by gohanlon** [reply]
@nathansalaun-sofia Great! And yes, by "attributed properties" I mean properties marked with attributes, which includes property wrappers, macros, and any other attribute semantics Swift supports now or may support in the future.

The key differences are:
1. `@_UncheckedMemberwiseInit` includes attributed properties in the generated initializer by default.
2. `@MemberwiseInit` excludes attributed properties by default and requires explicit inclusion via `@Init` or `@InitWrapper(type:)`.
3. With `@_UncheckedMemberwiseInit`, you can still use `@Init(.ignore)` on a specific attributed property if you need to exclude it from the initializer.

> **Note**
> Planned for 1.0, `@MemberwiseInit` will emit a diagnostic error for any attributed properties that don't have explicit configuration. See #19.

Omitting attributed properties by default in `@_UncheckedMemberwiseInit` is an attempt at consistency with its "unchecked" nature. The idea is that, similar to how it doesn't require `@Init(.public)` for properties with different access levels, it might not need additional annotations for attributed properties. However, the behavior is based more on an idea of consistency and anticipated developer expectations rather than practical application.

Thoughts?

**Reply by nathansalaun-sofia** [reply]
It might make for a heavier and boilerplate API but maybe we could specify what we don't want to be checked? Something like 

```swift
@_UncheckedMemberWiseInit(.public, ignoring: .visibility)
@_UncheckedMemberWiseInit(.public, ignoring: .attributedProperties)
```

I'm not quite satisfied with the naming but you get the idea

**Reply by gohanlon** [reply]
Full disclosure: I'm feeling rather far afield on this!

I don't personally have any use cases for `@ _UncheckedMemberwiseInit`. And, I seldomly apply `@MemberwiseInit` to structs that make use of attributed properties. This is largely due to an unresolved Swift Macros bug (see #26) that breaks `#Preview` functionality. Once this bug is addressed, I anticipate using `@MemberwiseInit` extensively with `SwiftUI.View`s, which almost always use attributed properties.

I don't think it's possible to create well-designed APIs without being informed by actual use, which I personally lack. If you have any real uses to share that concern MemberwiseInit's macros treatment of attributed properties, I'd be ecstatic. That said:

If real-world usage reveals that the distinctions between `@_UncheckedMemberwiseInit` and `@MemberwiseInit` are too nuanced or unclear to users, we should then—and only then—revisit the possibility of unifying these macros under a single, cohesive API.

Consider `@MembewiseInit`'s current signature:

```swift
@attached(member, names: named(init))
public macro MemberwiseInit(
  _ accessLevel: AccessLevelConfig,
  _deunderscoreParameters: Bool? = nil,
  _optionalsDefaultNil: Bool? = nil
) = …

@attached(member, names: named(init))
public macro MemberwiseInit(
  _deunderscoreParameters: Bool? = nil,
  _optionalsDefaultNil: Bool? = nil
) = …
```

Now, let's imagine a new signature that unifies the two macros:

```swift
public enum MemberwiseInitOptions {
  case uncheckedAccessLevels
  case includeAttributedProperties
  case _deunderscoreParameters
  case _optionalsDefaultNil
}

@attached(member, names: named(init))
public macro MemberwiseInit(
  _ accessLevel: AccessLevelConfig = .internal,
) = #externalMacro(
  module: "MemberwiseInitMacros",
  type: "MemberwiseInitMacro"
)

@attached(member, names: named(init))
public macro MemberwiseInit(
  _ accessLevel: AccessLevelConfig = .internal,
  options: MemberwiseInitOptions...
) = #externalMacro(
  module: "MemberwiseInitMacros",
  type: "MemberwiseInitMacro"
)
```

This design unifies all configuration options into a single enum, simplifying the macro's signature and plays well with Xcode's autocomplete. Here are some usage examples:

```swift
@MemberwiseInit(.public) // Default behavior
struct RegularStruct { ... }

@MemberwiseInit(.public, options: .uncheckedAccessLevels, .includeAttributedProperties)
struct FullyUncheckedStruct { ... }

@MemberwiseInit(.public, options: .uncheckedAccessLevels, ._deunderscoreParameters)
struct CustomStruct { ... }

@MemberwiseInit(.public, options: .includeAttributedProperties, ._optionalsDefaultNil)
struct AnotherCustomStruct { ... }
```

While this design shows promise, several concerns warrant careful consideration:

1. Backward Compatibility:
   - How would this new design affect existing codebases using the current `@MemberwiseInit`?
   - Implementation Strategy: We could introduce this in a minor release as an additive feature, while deprecating existing usages and providing fix-its for migration.
   - Caveat: This approach, while sound, represents a non-trivial development effort.

2. Option Combinations:
   - Current State: All existing options can be combined arbitrarily.
   - Future Considerations: As we add new options, some combinations may become invalid or undesirable.
   - Challenge: How can we design the API to gracefully handle potentially incompatible option combinations in the future?
   - Note: While we may not have a definitive answer now, exploring this potential issue upfront could inform our design decisions.

3. Global Configuration vs. Property-Level Safety:
   - Discoverability: Setting behavior "globally" for the struct is more discoverable in the new design, and that may not be a good thing.
   - Current Approach: Fix-its currently guide developers to think about each property individually and configure them deliberately when issues arise (e.g., properties with more restrictive access levels or attributed properties).
   - Potential Risk: The new usage may encourage developers to prefer less safe "global" configurations that effectively disable safety checks for the entire struct.
   - Swift Idioms: This shift away from property-level safety checks could be seen as less idiomatic to Swift, which typically favors fine-grained control and safety.
   - Design Question: How can we balance the ease of use provided by global options with the safety benefits of property-specific configurations? Perhaps we don't fully unify the two macros, and only offer `.uncheckedAccessLevels` and `.includeAttributedProperties` as options for `@_UncheckedMemberwiseInit`?

**Reply by nathansalaun-sofia** [reply]
I like the unified design with the `options` parameter. It seems to be the more expressive of all variants, and it allows extensions to add more options in the future without changing the macro itself and its usage.

Here's my opinion on some concerns that you raised:

> How would this new design affect existing codebases using the current @MemberwiseInit?

The new `options` parameter is optional. The default value must match the current behaviour of `@MemberwiseInit`.

> Implementation Strategy: We could introduce this in a minor release as an additive feature, while deprecating existing usages and providing fix-its for migration.

Why deprecate anything if the new parameter is optional and backwards compatible?

> Challenge: How can we design the API to gracefully handle potentially incompatible option combinations in the future?

A simple compromise would be to have the macro provide compile-time errors to indicate incompatible option combinations.

> Swift Idioms: This shift away from property-level safety checks could be seen as less idiomatic to Swift, which typically favors fine-grained control and safety.

I agree, however (at least in my case) this is a deliberate choice / compromise. We have a lot of "config" structs with tenths of properties and we _do not care_ about individual access control, we just want the structs to be constructible from everywhere without polluting the code too much.

**Reply by gohanlon** [reply]
> Why deprecate anything if the new parameter is optional and backwards compatible?

While the new parameter is optional and backwards compatible, it would introduce a new way to express the same functionality. For example:

```swift
@MemberwiseInit(.public, _optionalsDefaultNil: true)
```

becomes:

```swift
@MemberwiseInit(.public, options: ._optionalsDefaultNil)
```

Here's why I'd deprecate the older syntax in favor of a single, canonical usage:

- Reduced cognitive overhead: One syntax to learn and apply.
- Cleaner IDE autocomplete suggestions.
- Improved code consistency and readability.
- Simpler, more focused API documentation.
- Simplified future API evolution.

_Note: While `_optionalsDefaultNil` is currently an experimental configuration, I'm planning to include it formally in 1.0, with a simpler behavior (see #7)._

> … We have a lot of "config" structs with tenths of properties and we do not care about individual access control, we just want the structs to be constructible from everywhere without polluting the code too much.

You definitely made the case for `@_UncheckedMemberwiseInit`! 😃 As the sub-thread is rooted in its behavior of including attributed properties: Do your config structs happen to make use of attributed properties?

**Reply by hedukas** [reply]
This looks great! I also quite like the options parameter we could pass into the macro. It does make the call sites much cleaner and I think readable. Though yes, all the possible combinations that could arise are interesting. I'm guessing it would require fixits per combination? 

> Do your config structs happen to make use of attributed properties?

Very very little. I can only find a few. Adding the `Ignore` and `Default` attributes that come with this macro are the main ones used.

From what I tested on our codebase, it is very useful and reduced the amount of annotations and boilerplate code needed.

**Reply by nathansalaun-sofia** [reply]
> Do your config structs happen to make use of attributed properties?

Mine don't. We don't use MemberwiseInit for SwiftUI views but we may need to in the future, and attributed properties may come in play then.

Now that I think of it, SwiftUI views are a perfect example for `@_UncheckedMemberwiseInit(.public)`:
- you want the struct to be constructible from outside the module while exposing all properties in the initializer
- you _never_ need to (and even can) access the properties from outside the module once the structs are created (because of how the SwiftUI DSL works), so marking them public is useless
- all properties are always accessed from inside the struct (inside `body` for example)
- however you may need to expose attributed properties (such as bindings) in the initializer

If using `@MemberwiseInit`, marking all properties as `public` works but is useless because you cannot and you don't want to access them anyway, even if they are public. Marking them as `@Init(.public)` works as well. In any case, whether it's `public` or `@Init(.public)`, IMO it pollutes the code with something that's in the end out of our control.

**Reply by gohanlon** [reply]
Thank you, again, for your thoughtful feedback. I really appreciate the time you’ve taken to explore the macro’s potential applications and limitations. Your insights have sparked a valuable discussion that’s pushing me to think more deeply about how MemberwiseInit can better serve different use cases within the Swift ecosystem.

Let me address your points in detail, as I think they touch on some fundamental aspects of MemberwiseInit’s design.

> Now that I think of it, SwiftUI views are a perfect example for `@_UncheckedMemberwiseInit(.public)`:

I actually think SwiftUI views present some challenges for `@_UncheckedMemberwiseInit`.

As you alluded to by mentioning `@Binding` properties, the intricacies of SwiftUI’s complex property wrapper and macro ecosystem make it impossible to avoid explicit configuration.

`@_UncheckedMemberwiseInit`’s default behavior of including attributed properties can lead to two types of issues:

1. Immediate compiler errors, which at least alert the developer to a problem.
2. More insidiously, initializers that appear correct but lead to subtle bugs. These can manifest at runtime, potentially long after the code has been written and in unexpected scenarios.

Consequently, even when using `@_UncheckedMemberwiseInit`, developers will often need to provide explicit configuration for many properties. This remains true regardless of whether the macro’s default behavior is changed to exclude attributed properties.

In essence, correct property initialization needs a lot of careful consideration and manual intervention when used with SwiftUI views. `@_UncheckedMemberwiseInit` makes it all to easy to overlook that careful consideration. For SwiftUI views, I think `@MemberwiseInit` is more compelling (especially once it’s changed to enforce explicit configuration of attributed properties rather than ignoring them, #19).

> - you want the struct to be constructible from outside the module while exposing all properties in the initializer

To elaborate on the challenges with SwiftUI views, let’s consider some specific examples:

- `@State` properties should only be included in the initializer if the developer deliberately wants the “spooky” behavior that imparts.
- `@Binding` properties need to be configured to initialize the wrapper itself rather than the wrapped type.
- `@Environment` properties should never be part of the initializer.
- Private properties with default values, like `private var count = 0`, are sometimes used for internal state management during a single body evaluation and shouldn’t be exposed in the public initializer.

So, a configuration-less approach to property initialization is problematic for SwiftUI views. And, because Swift Macros aren’t given any type information, MemberwiseInit can’t even know that `@Binding` is a `@SwiftUI.Binding`, for instance. (#19 proposes a compromise: emit an error on attributed properties that lack explicit configuration, and offer fix-its that do assume `@Binding` is a `@SwiftUI.Binding`. That discussion also includes a more exhaustive list of and initialization treatments for SwiftUI property attributes.)

> - you _never_ need to (and even can) access the properties from outside the module once the structs are created (because of how the SwiftUI DSL works), so marking them public is useless
> - all properties are always accessed from inside the struct (inside `body` for example)

I agree that these aspects fit well with deliberate use of `@_UncheckedMemberwiseInit`. But it’s worth noting that, while they are generally true for SwiftUI views and I’d argue are best practices, they’re not guarantees. And, I’ve come across exceptions that are well designed.

> If using `@MemberwiseInit`, marking all properties as `public` works but is useless because you cannot and you don’t want to access them anyway, even if they are public. Marking them as `@Init(.public)` works as well. In any case, whether it’s `public` or `@Init(.public)`, IMO it pollutes the code with something that’s in the end out of our control.

Except in specific cases, I think viewing explicit configuration as pollution misses the mark. In general, explicit, co-located property declaration and initialization behavior enhances clarity, and alleviates or eliminates potentially severe initialization bugs.

MemberwiseInit is intended to be a general tool for the Swift language. It can’t make assumptions about specific usage patterns or frameworks. With the addition of  `@_UncheckedMemberwiseInit`, MemberwiseInit can offer an alternative behavior that better serves some, common specific use cases. Perhaps in the future, another addition can better serve SwiftUI views, too.
